py4web/utils/auth.py

```diff
@@ -36,7 +37,75 @@
 [ ] Lock account after x failed login attempts.
 [ ] Force new password every x days.
 """
-
+BLCOKED_IP_MAX_COUNT=5
+# timeout for blocked IP: if BLOCKED_IP_TIMEOUT=0 blocking never expired
+BLOCKED_IP_TIMEOUT=5*60         # reenable blocked ip after this time in seconds
+FILE_BLCOKED_IP_COUNT="host_ip_blocked.txt"
+FILE_HOST_IP_DENY="host_ip_deny.txt"
+FILE_HOST_IP_ALLOW="host_ip_allow.txt"
+host_ip_allow=[]
+host_ip_deny=[]
+host_ip_blocked={}
+try:
+    with open(FILE_HOST_IP_ALLOW, "r") as fp:
+         host_ip_allow = json.load(fp)
+except:
+    pass
+try:
+    with open(FILE_HOST_IP_DENY, "r") as fp:
+        host_ip_deny = json.load(fp)
+except:
+    pass
+
+try:
+    with open(FILE_BLCOKED_IP_COUNT, "r") as fp:
+        host_ip_blocked = json.load(fp)
+except:
+    pass
+
+def get_timestamp(): # returns timestamp in seconds from a specific time in the past
+    return  int((datetime.datetime.utcnow()-datetime.datetime(1,1,1,1,1,1)).total_seconds())
+
+def register_failed_login_ip(): # called if error in credentials
+    client_ip=request.remote_addr
+    blocked_ip=None
+    if host_ip_blocked:
+        blocked_ip = host_ip_blocked[client_ip]
+    else: # start new registration with timestamp in seconds
+        host_ip_blocked[client_ip]={'count':0,'last_failed':get_timestamp()}
+        blocked_ip = host_ip_blocked[client_ip]
+    blocked_ip['last_failed']=get_timestamp()   # update with the last failed time
+    print(request.path)
+    if blocked_ip['count']< BLCOKED_IP_MAX_COUNT:
+        blocked_ip['count']+=1
+        with open(FILE_BLCOKED_IP_COUNT, "w") as fp:
+            json.dump(host_ip_blocked, fp)
+
+ # called after successfule credentials validation to check if not blocked on too many attempts
+ # or black or white listed
+def ip_allow():
+    client_ip=request.remote_addr
+    is_this_ip_allowd=True
+    if host_ip_blocked:
+        blocked_ip =host_ip_blocked[client_ip]
```

Comment:
This is nice. Even if reading from DB make sure all blocked ips are read at statup to minimize disk IO at every request.